[Tablet Orders] Design tweaks

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Partially addresses #12057

## Description
This PR addresses the design issues raised on a previous review here: https://github.com/woocommerce/woocommerce-ios/pull/12033#pullrequestreview-1891221396

## Changes:
- Added a divider to the `ProductSelectorView`, between the header (clear selection, filter, ... ) and the product list. There's an additional divider on the top, separating the navigation controls from the rest of the view. This seems to be added by default from iOS16+, I was unable to remove it so far, needs more research.
<img width="421" alt="Screenshot 2024-02-22 at 10 01 13" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/6cb910c7-93d2-4202-976f-b439fdc03e3b">

- Addressed the segmented controls when searching products, and the search bar size itself being truncated on smaller screens: We will use either a `.segmented` or a `.menu` control type based on the screen size class and a minimum screen width (currently set to 450, happy to change this as you see fit).

![Simulator Screenshot - iPhone 15 Plus - 2024-02-22 at 08 48 08](https://github.com/woocommerce/woocommerce-ios/assets/3812076/71462bb0-8538-42c7-9f1f-5d866c7e47f8)

- Adjusted the product selector header in split views, the combination of padding and using `.fixedSize()` for some of their components made the controls overflow the available view. There's some improvements that could be done here regarding spacing and padding on smaller screens, currently we allow the HStack to grow vertically if needed, rather than forcing a fixed size and cutting out button text.

| Before | After |
|--------|--------|
| <img width="585" alt="Screenshot 2024-02-22 at 10 04 06" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/601a24d3-585e-4965-b336-a2a16148b408"> | ![simulator_screenshot_C3FC513A-6B66-4F5B-B34F-1C0D1559B52C](https://github.com/woocommerce/woocommerce-ios/assets/3812076/3d3e97b2-42b2-4831-a9fd-649aa7ca04ba) | 

- Addressed using the wrong tint on dark mode for the scanner button icon.

<img width="195" alt="Screenshot 2024-02-22 at 10 06 32" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/af189c95-5402-486a-9243-f591edc0ec21">

## Testing instructions
- Switch `sideBySideViewForOrderForm ` to `true`
- Navigate to the Order > `+` and observe that the UI has been updated as above. In general, test with iPhone, iPad, and iPhone that supports split views in landscape mode looking for edge cases.

iamgabrielma committed:
Add divider between header and product selector

iamgabrielma committed:
Segmented or menu search based on size class

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/187">17.5</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Change pickerstyle based on view width

iamgabrielma committed:
Extract numbers to extension

iamgabrielma committed:
Adjust search bar bg color and padding

iamgabrielma committed:
Remove fixed size modifiers

The views marked with .fixedSize make them overflow the safe area when we use split views.

iamgabrielma committed:
Reduce padding between search and picker

iamgabrielma committed:
Remove unnecessary component padding

iamgabrielma committed:
remove foreground color override

joshheald commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift:
```diff
@@ -301,27 +302,28 @@ private extension ProductSelectorView {
-            .padding(.trailing)
-            .renderedIf(searchHeaderisBeingEdited)
         }
+        .frame(height: Constants.minimumRowHeight * scale)
```
Why do we need to set the height?

iamgabrielma replied:
Good question, it's because of the `GeometryReader`. If we do not enforce a desired height for the view, it will take all the available space vertically, so needs to be constrained. If we don't, depending on the view, it would look like this:

![Simulator Screenshot - iPhone 15 Plus - 2024-02-23 at 11 06 50](https://github.com/woocommerce/woocommerce-ios/assets/3812076/51249cf2-a798-4cbc-91b5-9b952d91d100)

I've documented this on 78e717fdc29cb65dddfc31d52359a0bb385e668a

joshheald replied:
It's possible to put the geometry reader on the background of the views you care about, which prevents it messing with their layout. It always feels a little clunky, but perhaps it would be the right choice here to avoid setting the height ourselves...

I would rather find a design which avoids the need for a geometry reader though, if possible.

iamgabrielma replied:
TIL about using the geometry reader on the background. I'll experiment a combination of this and creating a separate component as you advised (sort of AdaptiveStack) separately and see if we can remove the geometry reader.

joshheald commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift:
```diff
@@ -301,27 +302,28 @@ private extension ProductSelectorView {
-            .renderedIf(searchHeaderisBeingEdited)
         }
+        .frame(height: Constants.minimumRowHeight * scale)
+        .padding(.bottom, Constants.defaultPadding)
```
This should probably be removed, or at least smaller.

iamgabrielma replied:
If we remove it, the search bar on tablets will have no spacing, since renders at the bottom of the header: 

<img width="389" alt="Screenshot 2024-02-23 at 10 56 29" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/03b8a0f1-5b83-4fee-bfde-57c13c1c851e">

I moved it outside of the View, when we decide how to render the views based on feature flag and size class, so only will apply when we render it in this setup: 545fe4b83814e899f52875205614fa16f47c0426

<img width="388" alt="Screenshot 2024-02-23 at 10 59 25" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/147b2dc9-dd9f-48f3-97e1-05146690ee66">

joshheald replied:
That _seems_ like it would be better handled by padding around a VStack containing 
`productSelectorHeaderSearchRow` and `productSelectorHeaderTitleRow` – at the moment they share the stack with the rows, which means you can't do that, but perhaps if they had their own stack, the padding would be easier to apply.

iamgabrielma replied:
That was my first approach, but got messy very fast. I'll followup as with https://github.com/woocommerce/woocommerce-ios/pull/12062#discussion_r1500424844 , attempting to create an adaptive component and perhaps extract some more bits to their own `@ViewBuilders` for easier composition.

joshheald commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift:
```diff
@@ -301,27 +302,28 @@ private extension ProductSelectorView {
+                Picker(selection: $viewModel.productSearchFilter, label: EmptyView()) {
+                    ForEach(ProductSearchFilter.allCases, id: \.self) { option in Text(option.title) }
+                }
+                .if(horizontalSizeClass == .regular && geometry.size.width < Constants.headerSearchRowWidth) { $0.pickerStyle(.menu) }
+                .if(horizontalSizeClass == .compact || geometry.size.width > Constants.headerSearchRowWidth) { $0.pickerStyle(.segmented) }
```
Depending how much we perfect it, some of this could be usefully moved in to a new generic view similar to AdaptiveStack. But I think it might be easier to simplify the design so we're not cramming so much in, and then rely on default stack behaviour.

joshheald reviewed (approved):
Looks better, but I don't think we're there yet with the header. I've approved so we can keep moving on this, but please make sure the view looks correct with the feature flag off before merging.

1. Spacing in fullscreen/landscape 
![Annotated screenshot showing strange spacing of header elements, and padding below search bar](https://github.com/woocommerce/woocommerce-ios/assets/2472348/e26a4a4c-e2d9-4b9b-9e9c-916135540abf)

2. Alignment of the clear/filter buttons is sometimes strange:
![Annotated screenshot showing alignment of filter/search buttons](https://github.com/woocommerce/woocommerce-ios/assets/2472348/2c236aa0-76f2-4154-9c86-620621a35a9e)

3. Even in portrait fullscreen, the header is too cramped. Moving the title to its own line, or even replacing the `Add Product` nav title with `Select products`/`n products selected` might help here.
![Annotated screenshot showing portrait iPad cramped header](https://github.com/woocommerce/woocommerce-ios/assets/2472348/6bac2e1f-d5ca-4cb5-aab1-877a9f9db20d)

4. In modal on modal, we might want to use the menu too... at least sometimes?
![IMG_244072B34014-1](https://github.com/woocommerce/woocommerce-ios/assets/2472348/054bf2d2-0395-45a0-abad-faff26b23355)

5. It would be better if we could avoid changing the view with the feature flag off. I think I missed this on the previous PR.
![Screenshot with feature flag off - spacing and padding have changed](https://github.com/woocommerce/woocommerce-ios/assets/2472348/1a60c5b1-ee3f-4678-bacd-1c64df397390)

Scan button colour looks good, and I like the menu option 👍

iamgabrielma committed:
Apply search bar bottom padding conditionally

This padding only should be applied when the product selector search is on the bottom, which only happens on regular size classes

iamgabrielma committed:
Add documentation

iamgabrielma committed:
Use different picker style based only on view width

iamgabrielma committed:
Restore filter button fixedSize and padding

iamgabrielma committed:
Render product selection conditionally of width

iamgabrielma commented:
Thanks for the review @joshheald , I'll add comments here as I address the different items: 

> In modal on modal, we might want to use the menu too... at least sometimes?

Good catch, in this case I think we can avoid checking the size class entirely: If the view width is smaller than X, we use `.menu`, otherwise, we use `.segmented`. Addressed on https://github.com/woocommerce/woocommerce-ios/pull/12062/commits/46e5b428a12f994fb29cd5b6f83b49ff5d2a8d1b

> 1. Spacing in fullscreen/landscape
> 2. Alignment of the clear/filter buttons is sometimes strange:
> 3. Even in portrait fullscreen, the header is too cramped. Moving the title to its own line, or even replacing the Add Product nav title with Select products/n products selected might help here.

These are tricky, by using `.fixedSize()` they align perfectly...until there's not enough screen width to have a fixed size, in that case they just overflow. 

For the moment I've addressed it by restoring the fixed sizes and _not_ rendering the `n selected items` title if the screen is too small, which is exactly what we do on `modalOnModal` ( https://github.com/woocommerce/woocommerce-ios/pull/12062/commits/b39d2725e836db55fec3464323d3f2a24e5ea0ec , https://github.com/woocommerce/woocommerce-ios/pull/12062/commits/1e3599bba774a5f81f48c24abd0544d02b88edb2 ) this fixes multiple alignment issues, but I'll continue experimenting with it and see how can be improved (eg: bottom padding in some cases needs to be adjusted a bit):

**.regular, big screen**

![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2024-02-23 at 12 15 43](https://github.com/woocommerce/woocommerce-ios/assets/3812076/2d86329d-ce67-4974-8d00-eeffbbe22868)

**.regular, smaller screen**

![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2024-02-23 at 12 15 53](https://github.com/woocommerce/woocommerce-ios/assets/3812076/b0068b6a-9457-4ea2-bfaa-d07f96f59fc5)

**.compact, smallest screen**

![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2024-02-23 at 12 16 01](https://github.com/woocommerce/woocommerce-ios/assets/3812076/8ba296ca-ef13-4c2a-8b47-0d4144f69b2a)

iamgabrielma committed:
remove whitespace

iamgabrielma commented:
> It would be better if we could avoid changing the view with the feature flag off. I think I missed this on the previous PR.

True! By restoring the `.fixedSize()` property for buttons with https://github.com/woocommerce/woocommerce-ios/commit/1e3599bba774a5f81f48c24abd0544d02b88edb2 we're back to the previous design when the flag is turned off:

![simulator_screenshot_9DDF0C97-F51C-44EC-A006-EBE1FB0FE019](https://github.com/woocommerce/woocommerce-ios/assets/3812076/76875d6f-6cae-4854-a7de-a12442400e50)

joshheald commented:
> > It would be better if we could avoid changing the view with the feature flag off. I think I missed this on the previous PR.
> 
> True! By restoring the `.fixedSize()` property for buttons with [1e3599b](https://github.com/woocommerce/woocommerce-ios/commit/1e3599bba774a5f81f48c24abd0544d02b88edb2) we're back to the previous design when the flag is turned off:

Not critical, but there's still a difference I think; the background colour and dividers. If we don't have time to fix it, it's a minor thing.
![IMG_3C2C832BDCC4-1](https://github.com/woocommerce/woocommerce-ios/assets/2472348/b18a1bac-ebee-4ef8-b85f-b24a426fb70d)

iamgabrielma commented:
> Not critical, but there's still a difference I think; the background colour and dividers. If we don't have time to fix it, it's a minor thing.

Double check you've pulled the latest updates in the PR, this was addressed in one of the most recent commits, both with `.sideBySideViewForOrderForm` on or off uses the `.listBackground` colour:
![simulator_screenshot_51556412-6475-44E9-A4DC-6B14F55C6F7E](https://github.com/woocommerce/woocommerce-ios/assets/3812076/569886fe-b437-4ca7-ada6-696fb860db49)

joshheald commented:
> For the moment I've addressed it by restoring the fixed sizes and _not_ rendering the `n selected items` title if the screen is too small, which is exactly what we do on `modalOnModal` ( [b39d272](https://github.com/woocommerce/woocommerce-ios/commit/b39d2725e836db55fec3464323d3f2a24e5ea0ec) , [1e3599b](https://github.com/woocommerce/woocommerce-ios/commit/1e3599bba774a5f81f48c24abd0544d02b88edb2) ) this fixes multiple alignment issues, but I'll continue experimenting with it and see how can be improved (eg: bottom padding in some cases needs to be adjusted a bit):

Perhaps putting it in the nav title would work? Or giving it its own line.

From the designs, that title and the `Clear` button go together – we could potentially remove the word `Selection` to give more space, if it were (as in designs) only rendered as `3 products selected [Clear]` or `Select products`... WDYT?

iamgabrielma commented:
> Perhaps putting it in the nav title would work? Or giving it its own line.
> From the designs, that title and the Clear button go together – we could potentially remove the word Selection to give more space, if it were (as in designs) only rendered as 3 products selected [Clear] or Select products... WDYT?

That sounds good, TBH I missed the bit on the designs for iPhone where the selected product count appears in the navigation bar instead of between the buttons. I'll merge this for the time being and tackle that along the experimentation around a different adaptive component next week 🙇
